### PR TITLE
[Reviewer: RJW2] BGCF-20 Break out WeightedSelector into a separate class

### PR DESCRIPTION
**Pull request by benritchie**
Rich
Can you review this please? - its the change you suggested as part of the Load-balancing you reviewed.  
Sprout UT's have been run.  is there any other component likely to be affected by this?
Thanks

### CODE REVIEW

File: include/weightedselector.h [hunk 1]
```diff
@@ -0,0 +1,166 @@
+  /**
```
Comment (richardwhiuk) [inline include/weightedselector.h]: This line shouldn't be indented.

File: src/baseresolver.cpp [hunk 2]
```diff
@@ -177,7 +178,7 @@ void BaseResolver::srv_resolve(const std::string& srv_name,
       srvs.reserve(i->second.size());
 
       // Build a cumulative weighted tree for this priority level.
-      SRVWeightedSelector selector(i->second);
+      WeightedSelector<BaseResolver::SRV> selector(i->second);
```
Comment (richardwhiuk) [inline src/baseresolver.cpp]: I think this can just be `WeightedSelector<SRV>`.

File: include/weightedselector.h [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+#ifndef WEIGHTEDSELECTOR_H__
+#define WEIGHTEDSELECTOR_H__
+
+#include <vector>
+
+#include "utils.h"
+#include "baseresolver.h"
+
+  /// The WeightedSelector class is a class used to implement resource
```
Comment (richardwhiuk) [inline include/weightedselector.h]: These lines shouldn't be indented.

File: include/weightedselector.h [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ * as those licenses appear in the file LICENSE-OPENSSL.
+ */
+
+#ifndef WEIGHTEDSELECTOR_H__
+#define WEIGHTEDSELECTOR_H__
+
+#include <vector>
+
+#include "utils.h"
+#include "baseresolver.h"
```
Comment (richardwhiuk) [inline include/weightedselector.h]: I don't think you should need either of these imports.

**Comment by richardwhiuk** [comment]
This mostly looks fine - some stylistic comments.

The UTs for the BaseResolver are in https://github.com/Metaswitch/cpp-common-test - specifically this is effectively tested in `src/ut/baseresolver_test.cpp` - you should run those - `make full_test` against the updated `cpp-common code` from the base directory will run the tests under valgrind.

It might be nice to add some generic unit tests just to test the template class, but you haven't reduced our coverage here, so I'm happy if you don't feel you have time to do that.

I assume this is sufficient for you integrate this externally?

File: include/weightedselector.h (hunk 3 above)
Comment (richardwhiuk) [inline include/weightedselector.h]: Oh and 'The WeightSelector is a template class used to implement resource...' would be nicer so we aren't repeating class twice in a single sentence - currently it reads oddly.

File: include/weightedselector.h (hunk 4 above)
Comment (benritchie) [inline include/weightedselector.h]: I think utils.h is needed to get access to std::string
Agree the other isn't.  - removed

Comment (richardwhiuk) [inline include/weightedselector.h]: `#include <string>` then?

**Pull request merged**